ngraph/core/reference/include/ngraph/runtime/reference/adaptive_avg_pool.hpp

```diff
@@ -0,0 +1,165 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <cfenv>
+#include <cmath>
+#include <numeric>
+#include <stdexcept>
+#include <vector>
+
+#include "ngraph/axis_vector.hpp"
+#include "ngraph/coordinate_transform.hpp"
+#include "ngraph/shape.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            inline size_t window_start(size_t idx, size_t arg_shape, size_t out_shape)
+            {
+                return floor(static_cast<double>(idx * arg_shape) / out_shape);
+            }
+            inline size_t window_end(size_t idx, size_t arg_shape, size_t out_shape)
+            {
+                return ceil(static_cast<double>((idx + 1) * arg_shape) / out_shape);
+            }
+            template <typename T>
+            T avg(const T sum, size_t n)
+            {
+                if (n == 0)
+                {
+                    throw std::runtime_error("AdaptiveAvgPool elements == 0, must be non-zero");
```

Comment:
Please use NGRAPH_CHECK for such assertions. 
`NGRAPH_CHECK(n != 0, "AdaptiveAvgPool elements == 0, must be non-zero");`